### PR TITLE
fix(authenticator): Move `autoSignIn` logic out of `signUp` actor

### LINKED ISSUE #1660: Unhandled Authenticator Route - Please open an issue: null
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Authenticator

### How is your app built?

Create React App

### Please describe your bug.

When attempting to create a user in an existing Cognito User Pool, I am receiving the error: 

```
'Unhandled Authenicator route – please open an issue: '
```

And redirected to the login page instead of being presented the appropriate MFA screen

### What's the expected behaviour?

Am expecting to be presented the screen to setup a User's MFA

### Help us reproduce the bug!

To make sure the issue wasn't with my application, I started a new app with `npx create-react-app my-app ` using node `14.19.0`. 

Next, I followed the getting started guide here: https://ui.docs.amplify.aws/getting-started/installation?platform=react as well as the Authenticator backend setup guide here: https://docs.amplify.aws/lib/auth/start/q/platform/js/#re-use-existing-authentication-resource. 

Next, I simply tried to create a user with the Create Account tab in the Authenticator component after running with `npm start`. After getting my confirmation code via Email and entering it, it returns to the login screen with the error in the console.

### Code Snippet

Here is the package.json I ended up with:

```
{
  "name": "my-app",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@aws-amplify/ui-react": "^2.15.1",
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^12.1.4",
    "@testing-library/user-event": "^13.5.0",
    "aws-amplify": "^4.3.19",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "react-scripts": "5.0.0",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

And  here was my App.jsx: 
```
import logo from './logo.svg';
import './App.css';
import { AmplifyProvider } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
import { Amplify } from 'aws-amplify';
import Auth from '@aws-amplify/auth';
import { Authenticator } from '@aws-amplify/ui-react';

Amplify.configure({
  Auth: {
    region: 'us-east-1',
    userPoolId: '<my-id>',
    userPoolWebClientId: '<my-other-id>',
    oauth: {
        options: {
            AdvancedSecurityDataCollectionFlag: true,
        },
        redirectSignIn: '<redirect>',
        redirectSignOut: '<redirect>',
        responseType: 'code',
        scope: ['email', 'profile', 'openid', 'aws.cognito.signin.user.admin'],
    },
  },
  storage: sessionStorage,
  API: {
      endpoints: [
          {
              name: 'api',
              endpoint: '<my-endpoint>',
              custom_header: async () => {
                  try {
                      const session = await Auth.currentSession();
                      const token = session.getIdToken().getJwtToken();
                      return { Authorization: `Bearer ${token}` };
                  } catch (error) {
                      console.error('User session not present to set jwt token');
                  }
              },
          },
      ],
  },
});




function App() {
  return (
    <AmplifyProvider>
      <Authenticator loginMechanisms={['email']} signUpAttributes={[
          "address",
          "birthdate",
          "email",
          "family_name",
          "gender",
          "given_name",
          "locale",
          "middle_name",
          "name",
          "phone_number",
          "profile",
      ]}>
        {({ signOut, user }) => (
          <main>
            <h1>Hello {user.username}</h1>
            <button onClick={signOut}>Sign out</button>
          </main>
        )}
      </Authenticator>
    </AmplifyProvider>
  );
}

export default App;
```

### Additional information and screenshots

Network tab showed two requests after confirming with the code. All returned 200s:

### Confirmation:
payload
```
{
  "ClientId": "********",
  "ConfirmationCode": "******",
  "Username": "***********@gmail.com",
  "ForceAliasCreation": true
}
```
response:
```
{}
```

### Number 2
payload:
```
{
  "AuthFlow": "USER_SRP_AUTH",
  "ClientId": "******",
  "AuthParameters": {
    "USERNAME": "*****@gmail.com",
    "SRP_A": "******"
  },
  "ClientMetadata": {}
}
```

Response:
```
{"ChallengeName":"PASSWORD_VERIFIER","ChallengeParameters":{"SALT":"***","USER_ID_FOR_SRP":"***"}}
```

### Number 3
payload:
```
{ChallengeName: "PASSWORD_VERIFIER", ClientId: "*********",…}
ChallengeName: "PASSWORD_VERIFIER"
ChallengeResponses: {USERNAME: "***********",…}
PASSWORD_CLAIM_SECRET_BLOCK: "*******"
PASSWORD_CLAIM_SIGNATURE: "**********"
TIMESTAMP: "Mon Apr 11 05:29:42 UTC 2022"
USERNAME: "***********"
ClientId: "************"
ClientMetadata: {}
```

Response:
```
{"ChallengeName":"MFA_SETUP","ChallengeParameters":{"MFAS_CAN_SETUP":"[\"SMS_MFA\",\"SOFTWARE_TOKEN_MFA\"]"},"Session":"**********"}
```

Cognito Pool Configs:
```
"EnvironmentUserPool": {
            "Type" : "AWS::Cognito::UserPool",
            "Properties" : {                
                "UserPoolName" :  {
                    "Fn::Sub": "asdf-${Environment}-users"
                },
                "UsernameAttributes": [
                    "email"
                ],
                "UsernameConfiguration": {
                    "CaseSensitive": "True"
                },
                "EmailConfiguration": {
                    "EmailSendingAccount" : "DEVELOPER",
                    "From" : "asdf@asdf.org",
                    "SourceArn" : "*****"
                },
                "Policies": {
                    "PasswordPolicy": {
                        "MinimumLength" : 14,
                        "RequireLowercase" : true,
                        "RequireNumbers" : true,
                        "RequireSymbols" : true,
                        "RequireUppercase" : true
                    }                    
                },
                "MfaConfiguration": "ON",
                "EnabledMfas": [
                    "SMS_MFA",
                    "SOFTWARE_TOKEN_MFA"
                ],
                "SmsConfiguration": {
                    "ExternalId": "*****",
                    "SnsCallerArn": {"Fn::GetAtt" : ["SmsSendingRole", "Arn"] },
                    "SnsRegion": {  "Ref" : "AWS::Region" }
                },
                "AutoVerifiedAttributes": [ "email" ]           
            }
        },
        "LocalUserPoolClient": {
            "Type" : "AWS::Cognito::UserPoolClient",
            "Properties" : {
                "AccessTokenValidity" : 60,
                "AllowedOAuthFlows" : [
                    "code"
                ],
                "AllowedOAuthFlowsUserPoolClient" : true,
                "AllowedOAuthScopes" : ["phone", "email", "openid", "profile", "aws.cognito.signin.user.admin"],
                "CallbackURLs" : [ 
                    { "Fn::FindInMap" : ["CallbackUrl", "local", "callbackUrl"] }
                 ],
                "ClientName" : {
                    "Fn::Sub": "local-app-client"
                },
                "ExplicitAuthFlows" : [ "ALLOW_ADMIN_USER_PASSWORD_AUTH",  "ALLOW_CUSTOM_AUTH", "ALLOW_USER_SRP_AUTH", "ALLOW_REFRESH_TOKEN_AUTH" ],
                "IdTokenValidity" : 60,
                "LogoutURLs" : [ 
                    { "Fn::FindInMap" : ["CallbackUrl", "local", "callbackUrl"] }
                 ],                
                "RefreshTokenValidity" : 1,
                "TokenValidityUnits" : {
                    "AccessToken" : "minutes",
                    "IdToken" : "minutes",
                    "RefreshToken" : "days"
                },
                "ReadAttributes": [
                    "address",
                    "birthdate",
                    "email",
                    "email_verified",
                    "family_name",
                    "gender",
                    "given_name",
                    "locale",
                    "middle_name",
                    "name",
                    "nickname",
                    "phone_number",
                    "phone_number_verified",
                    "picture",
                    "preferred_username",
                    "profile",
                    "updated_at",
                    "website",
                    "zoneinfo"
                ],
                "WriteAttributes": [
                    "address",
                    "birthdate",
                    "email",
                    "family_name",
                    "gender",
                    "given_name",
                    "locale",
                    "middle_name",
                    "name",
                    "nickname",
                    "phone_number",
                    "picture",
                    "preferred_username",
                    "profile",
                    "updated_at",
                    "website",
                    "zoneinfo"
                ],
                "SupportedIdentityProviders": [
                    "COGNITO"
                ],
                "PreventUserExistenceErrors": "ENABLED",
                "UserPoolId" :  { "Ref" : "EnvironmentUserPool" }
            }
        },
```

### PR DESCRIPTION
**Pull request by wlee221**
#### Description of changes

Previously, `autoSignIn` flow was part of the `signUp` actor. But `autoSignIn` is quite incomplete compared to what `signIn` actor does; `autoSignIn` only tries `Auth.signIn` and does not check for any auth challenges or edge cases like sms mfa, totp, etc.

Instead of having incomplete `autoSignIn` in `signUp` actor, this moves `autoSignIn` to `signIn` actor. This makes semantic sense, because `autoSignIn` is part of sign in flow anyway and `signUp` actor should only be dealing with sign up logic. 

#### Issue #, if available

Fixes #1660.

#### Description of how you validated changes

I confirmed manually that end-users are be automatically redirected to `confirmSignIn` after confirm sign up.

Also added new e2e test that goes through this process.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/ui/src/machines/authenticator/index.ts [hunk 1]
```diff
@@ -373,6 +374,7 @@ export function createAuthenticatorMachine() {
           event.data?.intent === 'confirmSignUp',
         shouldRedirectToResetPassword: (_, event) =>
           event.data?.intent === 'confirmPasswordReset',
+        shouldAutoSignIn: (_, event) => event.data?.intent === 'autoSignIn',
```
Comment (wlee221) [inline packages/ui/src/machines/authenticator/index.ts:377]: I realize after few months of my refactor that `intent` is a bad name. Really what `intent` is the `redirectTarget`. `{ intent: 'autoSignIn' }` means that authenticator should go to `autoSignIn` state next.

File: packages/react/src/components/Authenticator/Router/Router.tsx [hunk 2]
```diff
@@ -16,6 +16,7 @@ const getRouteComponent = (route: string) => {
     case 'authenticated':
     case 'idle':
     case 'setup':
+    case 'autoSignIn':
```
Comment (ErikCH) [inline packages/react/src/components/Authenticator/Router/Router.tsx:19]: We might need to add this case for Vue an Angular. I noticed an unusual visual regression where after you enter your code in the confirm sign up page, there is now a floating bar at the top of the page that appears for a second or two before the authenticated page appears. It's a bit distracting. It also appears at sign out.

Screen Shot 2022-05-19 at 8.53.22 AM

File: packages/ui/src/machines/authenticator/index.ts (hunk 1 above)
Comment (ErikCH) [reply packages/ui/src/machines/authenticator/index.ts:377]: 👍

File: packages/ui/src/machines/authenticator/actors/signIn.ts [hunk 3]
```diff
@@ -414,6 +473,9 @@ export function signInActor({ services }: SignInMachineOptions) {
 
           return validChallengeNames.includes(challengeName);
         },
+        shouldAutoSignIn: (context) => {
+          return context.intent && context.intent === 'autoSignIn';
```
Comment (slaymance) [inline packages/ui/src/machines/authenticator/actors/signIn.ts:477]:
```suggestion
          return context?.intent === 'autoSignIn';
```
Minor nit to make it consistent with `shouldAutoSignIn` in the `createAuthenticatorMachine`.

File: packages/ui/src/machines/authenticator/signUp.ts [hunk 4]
```diff
@@ -247,28 +241,17 @@ export function createSignUpMachine({ services }: SignUpMachineOptions) {
         setCodeDeliveryDetails,
         setUser,
         sendUpdate: sendUpdate(), // sendUpdate is a HOC
+        setAutoSignInIntent: assign({ intent: (_) => 'autoSignIn' }),
```
Comment (slaymance) [inline packages/ui/src/machines/authenticator/signUp.ts:244]: What's the reasoning for setting the `intent` property to be a function here? I don't see anywhere that we call `intent()` as a function, so maybe I'm misunderstanding something about how the `assign` call is used here.

File: packages/react/src/components/Authenticator/Router/Router.tsx (hunk 2 above)
Comment (wlee221) [reply packages/react/src/components/Authenticator/Router/Router.tsx:19]: Good catch! That's been frustrating to me. I'll create a new PR that targets this branch.

File: packages/ui/src/machines/authenticator/actors/signIn.ts (hunk 3 above)
Comment (wlee221) [reply packages/ui/src/machines/authenticator/actors/signIn.ts:477]: Ooh, good point!

File: packages/ui/src/machines/authenticator/signUp.ts (hunk 4 above)
Comment (wlee221) [reply packages/ui/src/machines/authenticator/signUp.ts:244]: Yeah, it's a function used by `assign` to set the value of `intent` to `"autoSignIn"`. More details on their [docs](https://xstate.js.org/docs/guides/context.html#assign-action).

You see this pattern in https://github.com/aws-amplify/amplify-ui/blob/main/packages/ui/src/machines/authenticator/actions.ts (with @eddiekeller's great comments!)

Comment (slaymance) [reply packages/ui/src/machines/authenticator/signUp.ts:244]: Gotcha, so this specifically gives us reusability for our actions. Makes sense!

**Review by slaymance — APPROVED** [review]

**Review by ErikCH — APPROVED** [review]
Just had the comment about fixing the floating line, during sign up. Otherwise looks good.

File: packages/react/src/components/Authenticator/Router/Router.tsx (hunk 2 above)
Comment (wlee221) [reply packages/react/src/components/Authenticator/Router/Router.tsx:19]: #1937 up!

**Pull request merged**